docs/roslyn_best_practices.md

```diff
@@ -0,0 +1,181 @@
+
+### 8. Treat the Code Fixer as a single unit of work
+Code fixers are applied as a one-step change and their order is not guaranteed. If the code you’re adding requires a namespace import, then your code fixer should perform that change if necessary. Do not assume that another code fixer will handle adding the correct namespace.
+
+## Considerations specific to upgrade-assistant
+
+There are many stopping points on the journey from .NET Framework to .NET latest and even more scenarios specific to our customers. Because of this, `upgrade-assistant` does not always assume that the code can be compiled. This section highlights some considerations specific to the "work in progress" state that all projects will pass through as they are upgraded.
+
+### 1. Do not assume the code will compile
+`upgrade-assistant` will make sweeping changes across the solution. These sweeping changes often result in manual changes that must be made after running upgrade assistant before the code will compile.
+
+As an example, `upgrade-assistant` will upgrade NuGet packages across major versions. `upgrade-assistant` does not evaluate if the newer package contains breaking changes.
+
+### 2. Symbols may not be resolvable
+`upgrade-assistant` will add, upgrade, and remove package references during the upgrade process. If your analyzer is looking for a specific symbol then you may need to [AddMetadataReference](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.project.addmetadatareference) to ensure the symbol is available.
+
+### 3. Add reference assemblies instead of the assembly implementation
+Reference assemblies are a special type of assembly that have only the smallest amount of metadata needed to represent the library's public API surface.
+
+If you need to [AddMetadataReference](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.project.addmetadatareference) then use a reference assembly when possible.
+
+Examples include:
+* [Microsoft .NET Framework Reference Assemblies .NET 4.8](https://www.nuget.org/packages/Microsoft.NETFramework.ReferenceAssemblies.net48/)
+* [Microsoft .NET Framework Reference Assemblies .NET 4.7.2](https://www.nuget.org/packages/Microsoft.NETFramework.ReferenceAssemblies.net472/)
+
+## Tips and more resources
+
+### 1. Use the Syntax Visualizer
+https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/syntax-visualizer
+
+The Syntax Visualizer is a tool window that helps you inspect and explore syntax trees. It's an essential tool to understand the models for code you want to analyze. It's also a debugging aid when you develop your own applications using the .NET Compiler Platform (“Roslyn”) SDK.
+
+### 2. Look at examples
+* [Roslyn-analyzers](https://github.com/dotnet/roslyn-analyzers)
+* [Roslynator](https://github.com/JosefPihrt/Roslynator)
+
+### 3. Video Training Courses
+* [Youtube: Learn Roslyn Now](https://www.youtube.com/watch?v=wXXHd8gYqVg) by Josh Varty
```

Comment:
Do we have other courses that we would recommend?

This looked relevant but I have not finished this course:
https://www.pluralsight.com/courses/vs-2015-diagnostic-analyzers-first-look